File: frontend/src/api/genesis.ts
import axios from 'axios';

// Create axios instance with base URL
const api = axios.create({
    baseURL: '/api/genesis',
    headers: {
        'Content-Type': 'application/json',
    },
});

export const setApiProvider = (provider: string) => {
    api.defaults.headers.common['X-Model-Provider'] = provider;
};

export interface WorldBackground {
    era: string;
    rules: string[];
    society: string;
}

export interface GameScene {
    name: string;
    description: string;
    key_objects: string[];
    locations?: string[];
}

export interface TimeConfig {
    start_datetime: string;
    display_year?: string;
    day_length_real_sec: number;
}

export interface WorldBible {
    world_id: string;
    title?: string;
    player_objective?: string;
    background: WorldBackground;
    scene: GameScene;
    time_config: TimeConfig;
}

export interface GenesisChatResponse {
    response: string;
    is_ready_to_generate: boolean;
    suggested_world_setting?: {
        title?: string;
        background: WorldBackground;
        scene: GameScene;
    };
    is_ready_for_npc?: boolean;
    npc_requirements?: string;
    npc_count?: number;
    is_ready_for_quest?: boolean;
    quest_requirements?: string;
    quest_count?: number;
    pending_actions?: PendingAction[];
}

export interface PendingAction {
    type: string;
    data: any;
    label: string;
    style: string;
}

export interface NPC {
    id: string;
    profile: {
        name: string;
        age: number;
        gender: string;
        race: string;
        avatar_desc: string;
        occupation: string;
        avatar_url?: string;
    };
    dynamic: {
        personality_desc: string;
        values: string[];
        mood: string;
        current_location: string;
        current_action?: string;
    };
    quest_role?: {
        role: string;
        clue: string;
        motivation: string;
        attitude: string;
    };
    goals: {
        id: string;
        description: string;
        type: string;
        status: string;
        trigger_condition?: string;
    }[];
    skills: {
        name: string;
        description: string;
        level: number;
    }[];
    relationships: Record<string, any>;
    _incomplete?: boolean;  // 标记 NPC 生成失败
    _error?: string;  // 错误信息
    _imageError?: boolean; // 图片生成失败
}

export interface QuestCondition {
    type: 'affinity' | 'item' | 'location' | 'time' | 'state' | 'dialogue';
    params: Record<string, any>;
}

export interface QuestNode {
    id: string;
    description: string;
    target_npc_id: string;
    required_affinity: number;
    status: 'pending' | 'active' | 'completed' | 'locked' | 'failed';
    conditions?: QuestCondition[];
    type?: 'dialogue' | 'collect' | 'investigate' | 'wait' | 'choice';
}

export interface Quest {
    id: string;
    title: string;
    type: 'main' | 'side';
    description: string;
    nodes: QuestNode[];
    status: 'active' | 'completed' | 'failed';
    target_npc_id?: string;
}

export default {
    async startSession() {
        const response = await api.post<{ session_id: string; message: string }>('/start');
        return response.data;
    },

    async chat(sessionId: string, content: string, currentBible?: any, phase: string = 'world') {
        const response = await api.post<GenesisChatResponse>('/chat', {
            session_id: sessionId,
            content,
            current_bible: currentBible,
            phase
        });
        return response.data;
    },

    async confirmWorld(sessionId: string, worldSetting: any, chatHistory?: any[]) {
        const response = await api.post<{ status: string; world_bible: WorldBible; is_locked: boolean }>('/confirm_world', {
            session_id: sessionId,
            world_setting: worldSetting,
            chat_history: chatHistory
        });
        return response.data;
    },

    async generateNPCs(worldBible: WorldBible, count: number = 3, requirements?: string) {
        const response = await api.post<{ npcs: NPC[] }>('/generate_npcs', {
            world_bible: worldBible,
            count,
            requirements
        });
        return response.data;
    },

    async generateRoster(worldBible: WorldBible, count: number = 3, requirements?: string) {
        const response = await api.post<{ skeletons: NPC[]; new_locations: string[] }>('/generate_roster', {
            world_bible: worldBible,
            count,
            requirements
        });
        return response.data;
    },

    async generateQuests(worldBible: WorldBible, npcs: NPC[]) {
        const response = await api.post<{ quests: Quest[] }>('/generate_quests', {
            world_bible: worldBible,
            npcs
        });
        return response.data;
    },

    async enrichAssets(worldBible: WorldBible) {
        const response = await api.post<{ items: any[]; locations: any[] }>('/enrich_assets', {
            world_bible: worldBible
        });
        return response.data;
    },

    async generateMainQuest(worldBible: WorldBible, npcs: NPC[], skipEnrichment: boolean = false, requirements?: string) {
        const response = await api.post<{ quests: Quest[]; items?: any[]; locations?: any[] }>('/generate_main_quest', {
            world_bible: worldBible,
            npcs,
            skip_enrichment: skipEnrichment,
            requirements
        });
        return response.data;
    },

    async generateSideQuest(targetNpc: NPC, worldBible: WorldBible) {
        const response = await api.post<{ quests: Quest[]; items?: any[]; locations?: any[] }>('/generate_side_quest', {
            target_npc: targetNpc,
            world_bible: worldBible
        });
        return response.data;
    },

    async generateNPCDetails(skeleton: any, worldBible: WorldBible, rosterNames: string, requirements?: string) {
        const response = await api.post<NPC>('/generate_npc_details', {
            skeleton,
            world_bible: worldBible,
            roster_names: rosterNames,
            requirements
        });
        return response.data;
    },

    async regenerateNPC(npcId: string, worldId: string, requirements?: string) {
        const response = await api.post<{ status: string; npc: NPC }>(`/regenerate_npc/${npcId}`, {
            world_id: worldId,
            requirements
        });
        return response.data;
    },

    async generateNPCImage(npcId: string, worldId: string, stylePrompt?: string) {
        const response = await api.post<NPC>('/generate_npc_image', {
            world_id: worldId,
            npc_id: npcId,
            style_prompt: stylePrompt
        });
        return response.data;
    },

    async saveChat(worldId: string, messages: any[]) {
        const response = await api.post<{ status: string }>('/save_chat', {
            world_id: worldId,
            messages
        });
        return response.data;
    },

    async updateNPCProfile(worldId: string, npcId: string, avatarDesc: string) {
        const response = await api.post<NPC>('/update_npc_profile', {
            world_id: worldId,
            npc_id: npcId,
            avatar_desc: avatarDesc
        });
        return response.data;
    },

    async uploadAvatar(worldId: string, npcId: string, file: File) {
        const formData = new FormData();
        formData.append('world_id', worldId);
        formData.append('npc_id', npcId);
        formData.append('file', file);

        const response = await api.post<NPC>('/upload_avatar', formData, {
            headers: {
                'Content-Type': 'multipart/form-data'
            }
        });
        return response.data;
    },

    async executeAddNPC(worldId: string, count: number, requirements?: string) {
        const response = await api.post<{ status: string }>('/execute_add_npc', {
            world_id: worldId,
            count,
            requirements
        });
        return response.data;
    },

    async executeRegenerateNPC(worldId: string, targetName: string, instruction: string) {
        const response = await api.post<{ status: string }>('/execute_regenerate_npc', {
            world_id: worldId,
            target_name: targetName,
            instruction
        });
        return response.data;
    },

    async executeManageImages(worldId: string, action: string, targetName?: string) {
        const response = await api.post<{ status: string }>('/execute_manage_images', {
            world_id: worldId,
            action,
            target_name: targetName
        });
        return response.data;
    },

    async executeConfirmRoster(worldId: string) {
        const response = await api.post<{ status: string }>('/execute_confirm_roster', {
            world_id: worldId
        });
        return response.data;
    },

    async updateQuest(worldId: string, questId: string, title: string, description: string) {
        const response = await api.post('/update_quest', {
            world_id: worldId,
            quest_id: questId,
            title,
            description
        });
        return response.data;
    },

    async checkStatus(worldId: string) {
        const response = await api.post<{
            world_id: string;
            is_locked: boolean;
            current_phase: string;
            progress: Record<string, any>;
            message: string;
        }>('/check_status', {
            world_id: worldId
        });
        return response.data;
    },

    async executeConfirmQuestBlueprint(worldId: string) {
        const response = await api.post<{ status: string }>('/execute_confirm_quest_blueprint', {
            world_id: worldId
        });
        return response.data;
    },

    // ✅ 新增直接获取 intro 的接口
    async getIntro(worldId: string) {
        const response = await axios.get<{ intro: string }>(`/api/worlds/${worldId}/intro`);
        return response.data;
    }
};

File: backend/app/schemas/genesis.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any

class WorldBackground(BaseModel):
    era: str = Field(..., description="时代背景，如赛博朋克、中世纪、现代")
    rules: List[str] = Field(default=[], description="世界法则，如高科技低生活、魔法存在")
    society: str = Field(..., description="社会结构")

class GameScene(BaseModel):
    name: str = Field(..., description="场景名称")
    description: str = Field(..., description="场景详细描述")
    key_objects: List[str] = Field(default=[], description="关键物品列表")
    locations: List[str] = Field(default=["Main Area"], description="场景内的具体地点列表 (如: 吧台, 门口, VIP座)")

class TimeConfig(BaseModel):
    start_datetime: str = Field(
        default="2024-01-01 08:00", 
        description="游戏开始时间 (YYYY-MM-DD HH:MM)，由 Agent 根据世界氛围推断"
    )
    display_year: Optional[str] = Field(
        default=None,
        description="纪元标签，如'黑暗纪元5年'、'魔法元年'，None时显示真实年份"
    )
    day_length_real_sec: int = Field(default=3600, description="现实一小时等于游戏一天")

class WorldBible(BaseModel):
    world_id: str = Field(..., description="世界唯一ID")
    title: Optional[str] = Field(default=None, description="世界标题")
    player_objective: Optional[str] = Field(default=None, description="玩家在这个世界的主要目标")
    background: WorldBackground
    scene: GameScene
    time_config: TimeConfig

# API Request/Response Models

class GenesisChatRequest(BaseModel):
    session_id: str = Field(..., description="会话ID")
    content: str = Field(..., description="用户输入内容")
    current_bible: Optional[Dict] = Field(default=None, description="当前的 World Bible (可选)，用于修改设定时提供明确上下文")
    phase: Optional[str] = Field(default="world", description="当前创世阶段 (world | npc | quest)")

class PendingAction(BaseModel):
    type: str = Field(..., description="Action type: add_npc, regenerate_npc, manage_images")
    data: Dict[str, Any] = Field(default={}, description="Action parameters")
    label: str = Field(..., description="Button label")
    style: str = Field(default="primary", description="Button style: primary, secondary, danger")

class GenesisChatResponse(BaseModel):
    response: str = Field(..., description="播种者回复")
    is_ready_to_generate: bool = Field(default=False, description="是否信息充足，可以生成世界")
    suggested_world_setting: Optional[Dict] = Field(default=None, description="推测的世界设定草稿")
    is_ready_for_npc: bool = Field(default=False, description="是否准备好生成NPC")
    npc_requirements: Optional[str] = Field(default=None, description="NPC生成要求")
    npc_count: Optional[int] = Field(default=3, description="NPC生成数量")
    is_ready_for_quest: bool = Field(default=False, description="是否准备好生成任务")
    quest_requirements: Optional[str] = Field(default=None, description="任务生成要求")
    quest_count: Optional[int] = Field(default=None, description="任务数量")
    pending_actions: Optional[List[PendingAction]] = Field(default=None, description="待确认的操作提案")

class GenerateWorldRequest(BaseModel):
    session_id: str
    world_setting: Optional[dict] = None # Optional override
    chat_history: Optional[List[Dict]] = None # Save chat history

class GenerateWorldResponse(BaseModel):
    status: str
    world_bible: WorldBible
    is_locked: bool = False

class GenerateQuestsRequest(BaseModel):
    world_bible: Dict = Field(..., description="World Bible Data")
    npcs: List[Dict] = Field(..., description="List of NPCs")

class GenerateQuestsResponse(BaseModel):
    quests: List[Dict] = Field(..., description="Generated Quests")
    items: List[Dict] = Field(default=[], description="Updated Items")
    locations: List[Dict] = Field(default=[], description="Updated Locations")

class GenerateMainQuestRequest(BaseModel):
    world_bible: Dict = Field(..., description="World Bible Data")
    npcs: List[Dict] = Field(..., description="List of NPCs (for context)")
    skip_enrichment: bool = Field(default=False, description="Skip asset reset and enrichment (if done previously)")
    requirements: Optional[str] = Field(default=None, description="User's Main Quest Requirements")

class EnrichAssetsRequest(BaseModel):
    world_bible: Dict = Field(..., description="World Bible Data")

class EnrichAssetsResponse(BaseModel):
    items: List[Dict] = Field(default=[], description="Optimized Items")
    locations: List[Dict] = Field(default=[], description="Optimized Locations")

class GenerateSideQuestRequest(BaseModel):
    target_npc: Dict = Field(..., description="Target NPC")
    world_bible: Dict = Field(..., description="World Bible Data")

class WorldCreationStatus(BaseModel):
    world_setting_finalized: bool = Field(default=False, description="世界设定是否定稿")
    npc_setting_finalized: bool = Field(default=False, description="NPC设定是否定稿")
    quest_setting_finalized: bool = Field(default=False, description="任务蓝图是否定稿")
    ready_to_start: bool = Field(default=False, description="是否可以启动世界")
    current_phase: str = Field(..., description="当前所处阶段 (World Bible | NPC | Quest | Launch | Ready)")
    message: str = Field(..., description="给用户的指导建议信息")
    details: Dict = Field(default={}, description="详细统计信息 (npc_count, avatar_count, quest_count)")
    progress: Dict[str, Dict] = Field(default={}, description="各阶段详细进度")

class WorldStatusFile(BaseModel):
    # 核心状态
    current_phase: str = Field(default="world", description="当前聚焦的阶段 (world, npc, quest, launch, ready)")
    
    # 额外选项
    options: Dict[str, Any] = Field(
        default={
            "is_illustrated": False, # 是否开启立绘
            "language_style": "normal", # 语言风格
        },
        description="世界生成的额外选项"
    )

    # 详细进度
    progress: Dict[str, Dict] = Field(
        default={
            "world_bible": { "status": "pending", "updated_at": None },  # pending, finalizing, completed
            "npc_roster": { "status": "pending", "count": 0, "updated_at": None }, # pending, drafting, finalizing, completed
            "quest_blueprint": { "status": "pending", "count": 0, "updated_at": None },
            "launch": { "status": "pending", "count": 0, "updated_at": None }, # pending, processing, completed
            # Launch phase sub-tasks
            "intro": { "status": "pending", "content": None, "message": None, "updated_at": None },
            "quest_enrich": { "status": "pending", "message": None, "updated_at": None },
            "schedule": { "status": "pending", "current": 0, "total": 0, "message": None, "updated_at": None }
        },
        description="各阶段详细进度"
    )
    
    # 上下文记忆
    phase_summaries: Dict[str, str] = Field(
        default={},
        description="各阶段的文本摘要 (world_summary, npc_summary, quest_summary)"
    )
    last_user_intent: str = Field(default="", description="用户最近的意图")
    active_task: str = Field(default="", description="当前 Agent 的具体任务")
    last_updated: str = Field(default="", description="最后更新时间")
